Deb aarch64

Felixoid opened the pull request:
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Rename main to pull_request
- Add BuilderDebAarch64 to all workflows
- Use docker buildkit, save building cache to docker hub and reuse it
- Build x86_64 and arm64 docker images separately, then merge them together to a multi-architecture manifest
- Tune many docker images to being multi-architecture
- Use the images from the current PR/commit in the following dependent builds
- Upgrade mysql client in stateless-tests image
- Add functional tests for aarch64 for PR actions (forced green for a while)
- Add python typing to some scripts
- Add  docker buildkit to runners' init script
- Add func-tester-aarch64 runners
- Use `docker login --password-stdin` to not expose password on exception

Felixoid committed:
Rename main/CIGithubAction to pull_request/PullRequestCI

Felixoid committed:
Add BuilderDebAarch64 to other workflows

Felixoid committed:
Add functional tests for aarch64

alesapin commented on .github/workflows/pull_request.yml:
```diff
@@ -1,4 +1,4 @@
-name: CIGithubActions
+name: CIPullRequest
```
PullRequestCI

Felixoid committed:
Apply black to docker_images_check.py

Felixoid committed:
Add typing to docker_images_check.py and ignore dependencies errors

Felixoid committed:
Use buildx for docker images, improve typing

Felixoid committed:
Add prebuilding for docker buildx builder

Felixoid committed:
Fix dpkg-deb architecture-dependent artifact

Felixoid committed:
EXPERIMENT: move docker building to gravitons

Felixoid committed:
Build dependent images from the current version

Felixoid committed:
Use cache feature of docker buildx

Felixoid committed:
Don't build image for arm64 pvs-sturio

qoega commented on tests/ci/docker_images_check.py:
```diff
@@ -16,36 +16,52 @@
+        )
 
-    dockerhub_repo_name = 'yandex'
+    dockerhub_repo_name = "yandex"
```
Still "yandex"?

Felixoid replied:
It's automatic [black](https://github.com/psf/black/) formatting. The PR it far from be finished, and I prefer to focus here on aarch64.

Other way, it will be quite complex to review.

But sure, makes sense to edit docker/images.json and lines 53,54 too. Especially since there's a bug, probably

qoega replied:
Didn't we move to "ClickHouse" dockerhub?

Felixoid replied:
It's here https://github.com/ClickHouse/ClickHouse/pull/32911/files#diff-5b4ed5d9824e0bd86ffb7baa16bb1d35417ca8bff01970350455c88179b5af07R53-R54

Felixoid replied:
It's gone to arguments parser for good now

Felixoid committed:
Use mysql-client-8.0 from ubuntu original repo

Felixoid committed:
Improve integration tests image

- Order packages
- Add rpm2cpio package
- Clean cache
- Use apache CDN for zookeeper

Felixoid committed:
Use a library with proper architecture for mysql odbc

Felixoid committed:
Add necessary 'before' field for a dummy object

Felixoid committed:
Build docker amd64 and arm64 separately and merge together later

Felixoid committed:
Use architecture specific version of minio

Felixoid committed:
Apply black to metrics lambda app

Felixoid committed:
Add func-testers-aarch64 group to metrics

Felixoid committed:
Build arch-dependent images in every workflow

Felixoid committed:
Add necessary MLIR dependency for codebrowser

Felixoid commented:
It has broken MySQL odbc currently =\

Felixoid committed:
Apply black to status_helper.py

Felixoid committed:
Force success aarch stateless tests

Felixoid committed:
Use stdin for docker login to not expose a password

alesapin reviewed (approved):
Great!

Felixoid commented on docker/test/codebrowser/Dockerfile:
```diff
@@ -1,12 +1,13 @@
 RUN sed -i "s|http://archive.ubuntu.com|$apt_archive|g" /etc/apt/sources.list
 
-RUN apt-get update && apt-get --yes --allow-unauthenticated install clang-13 libllvm13 libclang-13-dev
+RUN apt-get update && apt-get --yes --allow-unauthenticated install clang-13 libllvm13 libclang-13-dev libmlir-13-dev
```
Docker build fails now w/o it. But I can for sure make it separately.

Felixoid commented on tests/ci/docker_manifests_merge.py:
```diff
@@ -0,0 +1,235 @@
+
+    changed_images = get_changed_images(check_sources(to_merge))
+
+    os.environ["DOCKER_CLI_EXPERIMENTAL"] = "enabled"
```
https://docs.docker.com/engine/reference/commandline/manifest/#description

It's the same effect, but via ENV

Felixoid commented on docker/test/integration/base/Dockerfile:
```diff
@@ -1,44 +1,56 @@
-RUN curl 'https://downloads.mysql.com/archives/get/p/10/file/mysql-connector-odbc-8.0.21-linux-glibc2.12-x86-64bit.tar.gz' --location --output 'mysql-connector.tar.gz' && tar -xzf mysql-connector.tar.gz && cd mysql-connector-odbc-8.0.21-linux-glibc2.12-x86-64bit/lib && mv * /usr/local/lib && ln -s /usr/local/lib/libmyodbc8a.so /usr/lib/x86_64-linux-gnu/odbc/libmyodbc.so
+# Install MySQL ODBC driver from RHEL rpm
+RUN arch=${TARGETARCH:-amd64} \
+  && case $arch in \
```
pure bash =)

Well, you are kinda right

alesapin commented:
Let's merge.

alesapin commented:
Style check is green.
